### PR TITLE
sql: make multi-dimensional subscripts PG-like

### PR DESCRIPTION
**Pull request by sploiselle**
Our current implementation of subscripts treats each successive subscript operation as a left associative binary function, i.e. each subscript applies to the return value of the subscript operation to the left.

However, this is not how PG does it: instead, their multi-dimensional slices must be treated as variadic args to a function whose first argument is the item to subscript. This must be the case because arrays support infinitely-deep subscripts on any array type, and always return a value of the inner type:

```
sean=# SELECT pg_typeof('{1}'::_int2);
 pg_typeof
------------
 smallint[]
(1 row)

sean=# SELECT pg_typeof(('{1}'::_int2)[1][1][1][1]);
 pg_typeof
-----------
 smallint
(1 row)
```

This PR brings subscripts more inline with PG's semantics, but with the convenient property that multi-dimensional subscripts on arrays simply always returns null.

All this said, the feature this _actually_ unblocks is `int2vector[]`, which will be backed by arrays, and for which left-associative binary subscripts produces non-PG-like results, i.e. the subsequent subscripts index into the in2vector rather than the outer array.

```
sean=# SELECT pg_typeof('{1}'::_int2vector);
  pg_typeof
--------------
 int2vector[]
(1 row)

sean=# SELECT pg_typeof(('{1}'::_int2vector)[1][1]);
 pg_typeof
------------
 int2vector
(1 row)

sean=# SELECT pg_typeof((('{1}'::_int2vector)[1])[1]);
 pg_typeof
-----------
 smallint
```

@jkosh44 No need to do a thorough review because this is in the weeds, but certainly review to your heart's content

### Motivation

This PR adds a known-desirable feature: requisite for #10024

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: doc/user/content/release-notes.md [hunk 1]
```diff
@@ -140,6 +140,18 @@ changes that have not yet been documented.
 - Add the `md5`, `sha224`, `sha256`, `sha384`, and `sha512` [cryptography
   functions](/sql/functions/#cryptography-func).
 
+- Support sequences of subscript operations on [`array`] values when
+  indexing/accessing individual elements (as opposed to taking slices/ranges of
+  values). However, because Materialize does not currently support
+  multi-dimensional arrays, subscripting more than once always returns _NULL_
```
Comment (benesch) [inline doc/user/content/release-notes.md:146]: We do support multidimensional arrays! You just have to make them through the ARRAY constructor, IIRC.

**Review by benesch — COMMENTED** [review]
Just a skim for now, more later!

File: doc/user/content/release-notes.md (hunk 1 above)
Comment (sploiselle) [reply doc/user/content/release-notes.md:146]: Bleh--you do recall correctly!

File: src/sql-parser/src/ast/defs/expr.rs [hunk 2]
```diff
@@ -180,7 +180,7 @@ pub enum Expr<T: AstInfo> {
     /// `<expr>[<expr>]`
```
Comment (jkosh44) [inline src/sql-parser/src/ast/defs/expr.rs:180]: If I'm understanding this correctly, you may want to update this comment to something like `<expr>[<expr>]([<expr>])*`

Comment (sploiselle) [reply src/sql-parser/src/ast/defs/expr.rs:180]: Good catch, ty

File: doc/user/content/sql/types/list.md [hunk 3]
```diff
@@ -237,10 +237,10 @@ Layered slices
 {{< /experimental >}}
 
 To slice on layers beyond the first, indicate the range you want to slice along
-each layer, separating the ranges with commas:
+each layer, separating the ranges with square brackets:
```
Comment (mjibson) [inline doc/user/content/sql/types/list.md:241]: Is this a backward incompat change that needs to be called out in release notes?

File: src/expr/src/scalar/func.rs [hunk 4]
```diff
@@ -5109,6 +5099,27 @@ where
     }
 }
 
+fn list_index<'a>(datums: &[Datum<'a>]) -> Datum<'a> {
+    let mut buf = datums[0];
+
+    for i in datums[1..].iter() {
+        let i = i.unwrap_int64();
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:5132]: Do we know that there's no NULL here?

File: test/sqllogictest/arrays.slt [hunk 5]
```diff
@@ -306,9 +306,10 @@ SELECT ARRAY[1, 2, 3][1.5 + 1.5]
 ----
 3
 
-# Error different than Cockroach.
-query error cannot subscript type text
+query T
 SELECT ARRAY['a', 'b', 'c'][4][2]
+----
+NULL
```
Comment (mjibson) [inline test/sqllogictest/arrays.slt:312]: Unless I missed something, I didn't see a test in this file that tested what happens with NULL in the subscript place. Do we need to add some of those?

File: src/expr/src/scalar/func.rs [hunk 6]
```diff
@@ -5189,7 +5189,14 @@ fn list_slice<'a>(datums: &[Datum<'a>], temp_storage: &'a RowArena) -> Datum<'a>
     }
 
     temp_storage.make_datum(|row| {
-        slice_and_descend(datums[0], &ranges, row);
+        let empty_results = slice_and_descend(datums[0], &ranges, row);
+        // Empty results return an empty list and not NULL.
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:5219]: What's the motivation for this change? Does it need to be described in release notes?

**Review by mjibson — APPROVED** [review]

File: doc/user/content/sql/types/list.md (hunk 3 above)
Comment (sploiselle) [reply doc/user/content/sql/types/list.md:241]: Experimental changes don't need to get called out in release notes bc they're allowed to be undocumented breaking behavior.

File: test/sqllogictest/arrays.slt (hunk 5 above)
Comment (sploiselle) [reply test/sqllogictest/arrays.slt:312]: Big derp on my end; will add those tests/handling throughout.

File: src/expr/src/scalar/func.rs (hunk 6 above)
Comment (sploiselle) [reply src/expr/src/scalar/func.rs:5219]: This makes things behave like PG array for `slice_and_descend`; this should get called out in release notes because 1D slicing was _not_ gated behind the experimental flag, which I forgot. ty for the catch.

**Review by benesch — CHANGES_REQUESTED** [review]
Mentioned on Slack, but wanted to register the concern here more formally, just in case! I'm not sold on stabilizing this repeated slice behavior for lists because it doesn't match languages with a list-like type, like Python, JavaScript, even Rust. My mental model is that the list type is intentionally meant to deviate from the PostgreSQL array type _so that_ it can match the semantics of Python lists, JavaScript arrays, and Rust slices. I contend that similarity should extend to repeated slice operations, in which each slice operation simply continues to slice the outermost layer of the list, without ever peeling back the layers of the list.

IIRC, the thinking behind the old comma-separated subscript syntax for lists was so that we could preserve the Python/JS/Rust-compatible repeated slicing behavior while experimenting with an alternative multi-layer subscript operator for lists that _does_ peel back the layers of the list. FWIW I'm not aware of any prior art for this operator; I've only ever seen multi-dimensional slicing on matrix array types. 

Please do let me know if I'm missing something!

File: test/sqllogictest/arrays.slt (hunk 5 above)
Comment (sploiselle) [reply test/sqllogictest/arrays.slt:312]: Ah, this is automatically/already handled:

https://github.com/MaterializeInc/materialize/blob/main/test/sqllogictest/list.slt#L430
https://github.com/MaterializeInc/materialize/blob/main/test/sqllogictest/list.slt#L435

Comment (mjibson) [reply test/sqllogictest/arrays.slt:312]: :+1:

**Comment by sploiselle** [comment]
superseded by #10705
